sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/pubsublite/internal/SubscriptionPartitionProcessorImplTest.java

```diff
@@ -103,64 +105,59 @@ public void setUp() {
                 .build());
-    processor.close();
-    verify(subscriber).stopAsync();
-    verify(subscriber).awaitTerminated();
+    order.verify(subscriber).stopAsync();
+    order.verify(subscriber).awaitTerminated();
   }
 
   @Test
   public void lifecycleFlowControlThrows() throws Exception {
     when(tracker.currentRestriction()).thenReturn(initialRange());
     doThrow(new CheckedApiException(Code.OUT_OF_RANGE)).when(subscriber).allowFlow(any());
-    assertThrows(CheckedApiException.class, () -> processor.start());
-  }
-
-  @Test
-  public void lifecycleSubscriberAwaitThrows() throws Exception {
-    when(tracker.currentRestriction()).thenReturn(initialRange());
-    processor.start();
-    doThrow(new CheckedApiException(Code.INTERNAL).underlying).when(subscriber).awaitTerminated();
-    assertThrows(ApiException.class, () -> processor.close());
-    verify(subscriber).stopAsync();
-    verify(subscriber).awaitTerminated();
+    assertThrows(ApiException.class, () -> processor.runFor(Duration.ZERO));
   }
 
   @Test
   public void subscriberFailureFails() throws Exception {
     when(tracker.currentRestriction()).thenReturn(initialRange());
-    processor.start();
-    subscriber.fail(new CheckedApiException(Code.OUT_OF_RANGE));
+    doAnswer(
+            (Answer<Void>)
+                args -> {
+                  subscriber.fail(new CheckedApiException(Code.OUT_OF_RANGE));
+                  return null;
+                })
+        .when(subscriber)
+        .awaitRunning();
     ApiException e =
         assertThrows(
-            // Longer wait is needed due to listener asynchrony.
-            ApiException.class, () -> processor.waitForCompletion(Duration.standardSeconds(1)));
+            // Longer wait is needed due to listener asynchrony, but should never wait this long.
+            ApiException.class, () -> processor.runFor(Duration.standardMinutes(2)));
     assertEquals(Code.OUT_OF_RANGE, e.getStatusCode().getCode());
   }
 
   @Test
   public void allowFlowFailureFails() throws Exception {
     when(tracker.currentRestriction()).thenReturn(initialRange());
-    processor.start();
     when(tracker.tryClaim(any())).thenReturn(true);
     doThrow(new CheckedApiException(Code.OUT_OF_RANGE)).when(subscriber).allowFlow(any());
     leakedConsumer.accept(ImmutableList.of(messageWithOffset(1)));
-    ApiException e =
-        assertThrows(ApiException.class, () -> processor.waitForCompletion(Duration.ZERO));
+    ApiException e = assertThrows(ApiException.class, () -> processor.runFor(Duration.ZERO));
     assertEquals(Code.OUT_OF_RANGE, e.getStatusCode().getCode());
   }
 
   @Test
   public void timeoutReturnsResume() {
-    assertEquals(ProcessContinuation.resume(), processor.waitForCompletion(Duration.millis(10)));
+    assertEquals(ProcessContinuation.resume(), processor.runFor(Duration.millis(10)));
     assertFalse(processor.lastClaimed().isPresent());
   }
 
```

Comment:
Please add a test that would cover the case where downstream processing takes longer then `runFor` to prevent a regression.

Reply:
Downstream processing cannot take longer than runFor, or the client is violating its contract to not call the callback after it is shutdown (which runFor now does)